homeassistant/components/konnected.py

```diff
@@ -0,0 +1,315 @@
+        current_actuator_config = self.status.get('actuators')
+        _LOGGER.debug('%s: desired actuator config: %s', self.device_id,
+                      desired_actuator_config)
+        _LOGGER.debug('%s: current actuator config: %s', self.device_id,
+                      current_actuator_config)
+
+        if (desired_sensor_configuration != current_sensor_configuration) or \
+                (current_actuator_config != desired_actuator_config):
+            _LOGGER.debug('pushing settings to device %s', self.device_id)
+            self.client.put_settings(
+                desired_sensor_configuration,
+                desired_actuator_config,
+                self.hass.data[DOMAIN].get(CONF_ACCESS_TOKEN),
+                self.hass.config.api.base_url + ENDPOINT_ROOT
+            )
+
+
+class KonnectedView(HomeAssistantView):
+    """View creates an endpoint to receive push updates from the device."""
+
+    url = UPDATE_ENDPOINT
+    extra_urls = [UPDATE_ENDPOINT + '/{pin_num}/{state}']
+    name = 'api:konnected'
+    requires_auth = False  # Uses access token from configuration
+
+    def __init__(self, auth_token):
+        """Initialize the view."""
+        self.auth_token = auth_token
+
+    async def put(self, request: Request, device_id,
+                  pin_num=None, state=None) -> Response:
+        """Receive a sensor update via PUT request and async set state."""
+        hass = request.app['hass']
+        data = hass.data[DOMAIN]
+
+        try:  # Konnected 2.2.0 and above supports JSON payloads
+            payload = await request.json()
+            pin_num = payload['pin']
+            state = payload['state']
+        except json.decoder.JSONDecodeError:
+            _LOGGER.warning(("Your Konnected device software may be out of "
+                             "date. Visit https://help.konnected.io for "
+                             "updating instructions."))
+
+        auth = request.headers.get(AUTHORIZATION, None)
+        if not hmac.compare_digest('Bearer {}'.format(self.auth_token), auth):
+            return self.json_message(
+                "unauthorized", status_code=HTTP_UNAUTHORIZED)
+        pin_num = int(pin_num)
+        state = bool(int(state))
+        device = data[CONF_DEVICES].get(device_id)
+        if device is None:
+            return self.json_message('unregistered device',
+                                     status_code=HTTP_BAD_REQUEST)
+        pin_data = device[CONF_BINARY_SENSORS].get(pin_num) or \
+            device[CONF_SWITCHES].get(pin_num)
+
+        if pin_data is None:
+            return self.json_message('unregistered sensor/actuator',
+                                     status_code=HTTP_BAD_REQUEST)
+        entity = pin_data.get('entity')
+        if entity is None:
+            return self.json_message('uninitialized sensor/actuator',
+                                     status_code=HTTP_INTERNAL_SERVER_ERROR)
+
+        await entity.async_set_state(state)
```

Comment:
Entities should usually not be accessed directly from outside their platform. That's against the design philosophy of home assistant. Instead, we have the dispatch helper for this purpose, to send an update from the component to platform entities.

It's ok to store a key that references a specific device or entity, eg the entity_id in a dict to be able to send the update to the correct target with the dispatcher. But we shouldn't store the entities themselves so that they are accessible from outside the platform.

Reply:
How do I get the entity_id to store in a dict? It's unclear to me where this is set or how to retrieve it. `self.entity_id` returns `None` from inside the entity class. I must be missing something here.

Reply:
You might not need the entity_id. You don't need it specifically to use the dispatcher. I just mentioned it as an example identifier of an entity.

There's two parts to the dispatcher, a connect function and a send function. First you *connect* a target function to a target key. Then you *send* an update to the target key which will call all connected target functions of the target key.

The target key can be anything that is specific enough for what you want to target. If you want to target a specific entity, the entity_id is probably good. You could also use the builtin `id` function on the entity instance to get a unique id of the entity instance, if you don't need the key to be persistent between restarts.

You want to connect the entity target method from the entity coroutine method `async_added_to_hass`. At this point, the entity_id will be available on the entity, ie it will be something other than `None`.

Reply:
I understand conceptually.
Could you point me to a component that uses this pattern? This would help me tremendously.